docs/howto/build-run.md

```diff
@@ -222,6 +226,73 @@ version number starts with "1.8", and with JDK 11, it starts with
+
+* What went wrong:
+Could not compile build file '/Users/chrisbobbe/dev/zulip-mobile/node_modules/@unimodules/react-native-adapter/android/build.gradle'.
+> startup failed:
+  General error during semantic analysis: Unsupported class file major version 57
+
+  java.lang.IllegalArgumentException: Unsupported class file major version 57
+        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:184)
+        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:166)
+        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:152)
+        at groovyjarjarasm.asm.ClassReader.<init>(ClassReader.java:273)
+[...]
+```
+
+This can sometimes happen if you're using JDK 13 to invoke the build
+command (e.g., when calling `react-native run-android`, or
+`tools/test android`, or
+`android/gradlew -p android :app:assembleDebug`). You can check the
+version by running `java -version`. It seems that upgrading to
+macOS 10.15 Catalina automatically upgrades Java to 13.
+
+Somehow, JDK 8 is getting involved, and trying to use classfiles that
+it doesn't understand because they were created with JDK 13. We're not
+sure how JDK 8 is getting involved; it still seems to be involved
+after locating and killing some Gradle-related processes with
+`ps auxwww | grep gradle` that were using a JDK 8 installation that's
+bundled with Android Studio (yours, if you have one, might be at
+"/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/jre/bin/java").
+
+A solution, if you can't find why JDK 8 is getting involved and
+prevent that, is to ensure those classfiles are also created with
+JDK 8. (JDK 11 might also work; we haven't tested this.)
+
+1. First, find a JDK installation. When you check a Java
+   version, note that the version number will start with "1.8" for
+   JDK 8.
+   - As noted above, you may have a JDK installation included in
+   Android Studio at a path like the one above. Check its version
+   by running, e.g.,
+   `"/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/jre/bin/java" -version`.
+   - Otherwise, on Ubuntu or Debian, you can install it with
+   `sudo apt install openjdk-8-jdk`.
+   - In general, you can [download it from Oracle][jdk-8-oracle-dl].
+2. Find a path to that installation to set `JAVA_HOME` to. (This is
+   used by `gradlew` to find the `java` command to run Gradle under.)
+   - If you used the installation included in Android Studio, take
+     that path minus the "/jre/bin/java" at the end, e.g.,
+     "/Applications/Android Studio.app/Contents/jre/jdk/Contents/Home".
+   - If you installed the Debian/Ubuntu package, then
+     `ls -l /usr/lib/jvm/` should list `java-8-openjdk-amd64` or
+     something similar to that.
+3. Using the path from step 2, run `JAVA_HOME=[that path]`. Then, in
+   the same terminal window, try building again. If it works, add
+   `export JAVA_HOME=[that path]` to your `~/.bashrc` or equivalent,
+   so you don't have to repeat it in the future.
```

Comment:
So the way this syntax works is a bit counterintuitive:

(a) `$ FOO=bar` just sets `FOO` to `bar` as a variable in the shell, so `"$FOO"` gets turned to `bar`; it doesn't affect the environment, for commands run from the shell.

(b) `$ FOO=bar some-command` runs the command with `FOO` set to `bar` in its environment. It doesn't affect the shell variable `FOO`, or the environment for future commands.

(c) `$ export FOO=bar` sets the shell variable `FOO` just like `FOO=bar`, but also "exports" it, meaning that the same key-value pair `FOO=bar` will be added to the environment for future commands executed.

I think I gave you a command like (b) to use in the course of debugging, something like `JAVA_HOME=... android/gradlew ...`. And (c) is what you'd want to do once debugging/investigation is complete. The form (a) is only really useful in shell programming :-)

[Docs](https://www.gnu.org/software/bash/manual/bash.html#Environment); quick demo:
```
# (a), within shell
$ ( FOO=bar; echo $FOO; )
bar
# (a), within command
$ ( FOO=bar; bash -c 'echo $FOO' )

# (b)
$ ( FOO=bar bash -c 'echo $FOO' )
bar
# (c)
$ ( export FOO=bar; bash -c 'echo $FOO' )
bar
```

Oh, one wrinkle, I guess: if the shell already has `FOO` marked as "exported", then changes to the shell variable's value (with form (a)) will apply to the environment for future commands executed. In particular if `FOO` was in the shell's own environment when it was itself started, then that will have caused it to mark it as exported, so form (a) works:

```shell
# (a), within a shell that got a different value in the environment
$ ( export FOO=bar; bash -c 'FOO=zzz; echo $FOO' )
zzz
```

I think I wouldn't want to rely on that, though -- when the value is intended to appear in the environment, better to be explicit with either (c) or (b).

Reply:
Oh, OK, thanks! This helps my understanding a lot.

https://github.com/zulip/zulip-mobile/commit/0e45fda87

> I think I gave you a command like (b) to use in the course of debugging, something like `JAVA_HOME=... android/gradlew ...`.

That's right. So, I'll change the instructions so they say to do that.

> And (c) is what you'd want to do once debugging/investigation is complete.

What do you think about the additional recommendation to put it in `~/.bashrc`? This idea (and some other details, like JDK 8 installation instructions) was from https://github.com/zulip/zulip-mobile/commit/0e45fda87.